Argo jupyter scheduler (#465)
nebari-dev/nebari-docs: docs/docs/tutorials/run-notebook-as-a-job.md

---
id: jupyter-scheduler
title: Submit notebook jobs with jupyter-scheduler
description: Use jupyter-scheduler to submit scheduled or unattended jobs
---

# Submit Notebook jobs with jupyter-scheduler

:::warning
This is a new feature still in beta so please [leave some feedback][feedback-link] and [report any issues][issues-link].

There is one known issue with the `Update Job Definition` and `Resume` job definition which is related to [Nebari-Workflow-Controller issue #18][workflow-controller-issue]. The current workaround for those who need to update (or pause) your job definitions, is simply to delete the current job definition and create a new one as and when needed.
:::

> A common need that many Nebari users have is to submit their notebooks to run as a script or to run on a predefined schedule. This is now possible with [Jupyter-Scheduler][jupyter-scheduler], a JupyterLab extension that has been expanded and integrated into Nebari. This also allows users to view the status of their jobs by visiting the `<nebari-domain>/argo` endpoint.

Reviewer: ```suggestion
Nebari users often require scheduling notebook runs. This is now achievable with [Jupyter-Scheduler][jupyter-scheduler], a JupyterLab extension that has been enhanced and integrated into Nebari. Additionally, users can monitor their job status by accessing the `<nebari-domain>/argo` endpoint.
```

:::info
Shutting down your JupyterLab server (and associated JupyterLab extensions) doesn't stop your running Jupyter-Scheduler jobs because the Nebari team extended Jupyter-Scheduler to use Argo-Workflows as it's backend. Details can be found in the [argo-jupyter-scheduler repo][argo-jupyter-scheduler].

:::

Jupyter-Scheduler is included by default in the base Nebari JupyterLab image and can be used with any conda-store environment available to the notebook author.

:::note
By default, only users in the `admin` or `developer` groups will have permission to create notebook jobs. For more information regarding users and groups, please visit [How to configure Keycloak][configure-keycloak]
:::

## Submitting a Notebook as a Jupyter-Scheduler Job

To submit your job, simply click the `Jupyter-Scheduler` icon on the top of your notebook toolbar.

![Jupyter-Scheduler UI - location of the icon on the notebook toolbar](/img/tutorials/jupyter-scheduler-icon.png)

> This will open the Jupyter-Scheduler UI. From here you can specify:
> 
> - the notebook **Job name**
> - the **Input file** to use (this will default to the file from which the icon was clicked)
> - the **Environment** in which to run the notebook
> - **Output formats**
> - environment variables (**Parameters** section), if needed

Reviewer: We can remove the details and simply say: 
```suggestion
This will open a new **Create Job** tab. Fill the form and click "Create" button to create a schedule. 
```

<div align="center">
  <img src="/img/tutorials/jupyter-scheduler-run-now.png" alt="Jupyter-Scheduler UI - layout of Jupyter-Scheduler UI." width="60%"/>
</div>

:::note
Include the [`papermill` package][papermill] in your conda-store environment.
:::

Once created, the status of the job and the notebook output can be viewed from the Jupyter-Scheduler Notebook Jobs UI:

![Jupyter-Scheduler UI - view the notebook job status](/img/tutorials/jupyter-scheduler-job-status.png)

The output of your job will be accessible via the download icon in the "Output files" column. After you download the output, new links will replace the icon and point to the location of the downloaded output.

> Click on the notebook job name to view more information about the job, such as:
> 
> - the **Job ID**
> - when the job was **Created**
> - the **Start time** and **End time**

Reviewer: Like wise.
```suggestion
Click on the notebook job name to view more information about the job. 
```
